ocaml/xapi/xapi_pusb.ml

```diff
@@ -31,6 +31,61 @@ let create ~__context ~uSB_group ~host ~other_config ~path
   debug "PUSB ref='%s' created" (Ref.string_of pusb);
   pusb
 
+let scan_start ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let usbs = Xapi_pusb_helpers.get_usbs ~__context in
+  let existing_pusbs = List.filter (fun (rf, rc) -> rc.API.pUSB_host = host) (Db.PUSB.get_all_records ~__context) in
+  let rec find_or_create cur = function
```

Comment:
I'd like to see this simplified as I believe this less complicated than it looks. The task is:

* Scan localhost for the set of *local_usb* devices 
* Get the set of *known_usb* devices from xapi
* *create* new USB devices that are *unknown* to xapi
* *remove* USB devices that known to xapi but don't exist on the local host.

This should be reflected in the functions and let-bindings being used. There is a small difficulty that how a USB device is identified by the scanning code and by xapi is slightly different. It might be worth introducing a small record type that captures the identity of a USB device unless there is already a convenient type.

```
module USB = struct
  type t = {
    ...  (* something here *)
  }
  let compare = Pervasives.compare
end
module USBSet = Set.Make(USB)
```

This would give you the ability to create sets of USB devices to compare them easily to find the new and obsolete sets of USB devices.

Reply:
Thanks @lindig . A USB device is identified by (path, vendor-id, product-id ,serial). So here you mean add these to type t , right?

Reply:
Hi @lindig , when we plan to use USBSet to do this, seems we need to get usb_diff  by using USBSet.diff  with current_usbs and existing_usbs, in which current_usbs can get by parsing json format data and exsting_usbs can get from the database (it need to transform the database data to USB.t). When we get usb_diff, we need to get usb_to_add and usb_to_del to update PUSB records. For usb_to_add,  we will create new PUSB objects from elt of usb_to_add. But when we want to destroy the obsolete PUSBs , we need to get corresponding PUSB objects from the usb_to_del Set which need compare USB.t with PUSB records in database.  
So I think introduce a USBSet to do this may not make the code easier.  How do you think about it. May be have an easier way to do this :)

Reply:
The problem that the data types returned by the scan code and the database are different always exists. The Set module could act as a common ground and allows to think in terms of sets rather than lists.

Maybe the benefits of using a set are not worth doing it. In any case I'd like to see an implementation in which the functions and names reflect the task at hand and not just does something. Feel free to propose a solution without using a set.